### PR TITLE
Suggest dereferencing when `Deref` is implemented.

### LINKED ISSUE #59819: Dereferencing `Deref` types other than references is missing suggestion
The `b = a` assignment below suggests dereferencing `a` in order to make the code compile. The same suggestion applies to `y = x`, but no suggestion is produced by the compiler. I think we should report this suggestion for all type mismatches that satisfy `GotType: Deref<Target = ExpectedType>`

```rust
struct Foo(i32);

impl std::ops::Deref for Foo {
    type Target = i32;
    fn deref(&self) -> &i32 {
        &self.0
    }
}

fn main() {
    let x = Foo(42);
    let y: i32 = x;
    let a = &42;
    let b: i32 = a;
}
```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=eaae70a2b1f34a317cc3dcb6b7481155))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0308]: mismatched types
  --> src/main.rs:12:18
   |
12 |     let y: i32 = x;
   |                  ^ expected i32, found struct `Foo`
   |
   = note: expected type `i32`
              found type `Foo`

error[E0308]: mismatched types
  --> src/main.rs:14:18
   |
14 |     let b: i32 = a;
   |                  ^
   |                  |
   |                  expected i32, found &{integer}
   |                  help: consider dereferencing the borrow: `*a`
   |
   = note: expected type `i32`
              found type `&{integer}`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.

```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #59819.

r? @oli-obk 
cc @estebank

### CODE REVIEW

File: src/test/ui/suggestions/issue-59819.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+impl std::ops::Deref for Foo {
+    type Target = i32;
+    fn deref(&self) -> &i32 {
+        &self.0
+    }
+}
+
+fn main() {
+    let x = Foo(42);
+    let y: i32 = x; //~ ERROR mismatched types
```
Comment (oli-obk) [inline src/test/ui/suggestions/issue-59819.rs]: will the same suggestion occur if `Target = String`?

Comment (davidtwco) [reply src/test/ui/suggestions/issue-59819.rs]: Do you mean this case?

```rust
struct Foo(String);

impl std::ops::Deref for Foo {
    type Target = String;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn main() {
    let x = Bar("bar".to_string());
    let y: i32 = x; // no suggestion
}
```

Or this case?

```rust
struct Foo(String);

impl std::ops::Deref for Foo {
    type Target = String;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

fn main() {
    let x = Bar("bar".to_string());
    let y: String = x; // a suggestion
}
```

Comment (davidtwco) [reply src/test/ui/suggestions/issue-59819.rs]: Oh, I see what you mean.

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -463,7 +463,37 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                        ),
+                        // `Deref::Output`
+                        item_def_id,
+                    },
+                    // `U`
+                    ty: expected,
+                }));
+                let obligation = traits::Obligation::new(self.misc(sp), self.param_env, predicate);
+                if self.infcx.predicate_may_hold(&obligation) {
+                    if let (Ok(code), true) = (cm.span_to_snippet(sp), sp == expr.span) {
```
[the remark this answers is not in the document]
Comment (davidtwco) [reply src/librustc_typeck/check/demand.rs]: Fixed.

File: src/test/ui/suggestions/issue-59819.rs (hunk 1 above)
Comment (davidtwco) [reply src/test/ui/suggestions/issue-59819.rs]: I think this should be fixed now.

File: src/librustc_typeck/check/demand.rs [hunk 3]
```diff
@@ -463,6 +463,57 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 }
             }
+            _ if sp == expr.span => {
+                // If neither type is a reference, then check for `Deref` implementations by
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]: Seems like a good place to refactor into function(s)?

File: src/librustc_typeck/check/demand.rs [hunk 4]
```diff
@@ -463,6 +463,57 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                            // `Copy`
+                            copy_trait,
+                            // `T`
+                            self.tcx.mk_substs_trait(
+                                expected,
+                                self.fresh_substs_for_item(sp, copy_trait),
+                            ),
+                        ),
+                    }));
+                    if self.infcx.predicate_may_hold(&traits::Obligation::new(
```
Comment (oli-obk) [inline src/librustc_typeck/check/demand.rs]: copy checks can be done simpler (see the preexisting `*` suggestions above in https://github.com/rust-lang/rust/pull/60155/files#diff-124df07081a1185b5a4d40cdcf0cf4ffR443)

**Comment by oli-obk** [comment]
General note: When I wrote the issue I hoped that we could deduplicate the logic on references with the new logic you wrote. References also implement `&T: Deref<Target = T>`, so it should hopefully work out of the box?

**Comment by davidtwco** [comment]
> General note: When I wrote the issue I hoped that we could deduplicate the logic on references with the new logic you wrote. References also implement `&T: Deref<Target = T>`, so it should hopefully work out of the box?

I just tried this and I don't think it does, when I check if `&{integer}: Deref<Target = i32>` then it returns `false`. I assume that this is because `&{integer}: Deref<Target = {integer}>` and not `i32` specifically but that's just a guess.

File: src/librustc_typeck/check/demand.rs (hunk 4 above)
Comment (davidtwco) [reply src/librustc_typeck/check/demand.rs]: Fixed.

**Comment by oli-obk** [comment]
> I just tried this and I don't think it does, when I check if &{integer}: Deref<Target = i32> then it returns false. I assume that this is because &{integer}: Deref<Target = {integer}> and not i32 specifically but that's just a guess.

Maybe we need to try something different then. Check whether the type implements `Deref`, get the `Target` type, try to coerce to the destination type.

**Comment by davidtwco** [comment]
> Maybe we need to try something different then. Check whether the type implements `Deref`, get the `Target` type, try to coerce to the destination type.

I have no idea what I was trying earlier, pushed a version that de-duplicates that logic.

File: src/test/ui/suggestions/issue-59819.stderr [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+              found type `&{integer}`
+
+error[E0308]: mismatched types
+  --> $DIR/issue-59819.rs:32:21
+   |
+LL |     let g: String = f;
+   |                     ^
+   |                     |
+   |                     expected struct `std::string::String`, found struct `Bar`
+   |                     help: try using a conversion method: `f.to_string()`
```
Comment (estebank) [inline src/test/ui/suggestions/issue-59819.stderr]: I'm in awe of this suggestion working 🤯

File: src/test/ui/suggestions/issue-59819.rs [hunk 6]
```diff
@@ -0,0 +1,33 @@
+#![allow(warnings)]
```
Comment (estebank) [inline src/test/ui/suggestions/issue-59819.rs]: Mark the test `// run-rustfix` to make sure we don't start giving incorrect suggestions down the line.

Comment (davidtwco) [reply src/test/ui/suggestions/issue-59819.rs]: Fixed. I had this originally but removed it when I added the test case to make sure the suggestion wouldn't trigger for `String`, not realising that had a suggestion that meant rustfix would still work.

**Comment by oli-obk** [comment]
@bors r+

This is absolutely awesome!

**Pull request merged**